test/e2e/main_test.go

```diff
@@ -0,0 +1,56 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"reflect"
+	"runtime"
+	"testing"
+
+	"github.com/knative/eventing/test"
+)
+
+// channelTestMap indicates which test cases we want to run for a given CCP.
+var channelTestMap = map[string][]func(t *testing.T){
+	"in-memory-channel": []func(t *testing.T){
+		TestSingleBinaryEvent,
+		TestSingleStructuredEvent,
+		TestEventTransformation,
+		TestChannelChain,
+		TestDefaultBrokerWithManyTriggers,
+	},
+}
+
+func TestMain(t *testing.T) {
+	// if the main test is not indicated to be run, skip it directly.
+	if !test.EventingFlags.RunFromMain {
+		t.Skip()
+	}
+
+	var provisioners = test.EventingFlags.Provisioners
+	for _, provisioner := range provisioners {
+		// set the current provisioner that is used to run the test cases
+		ClusterChannelProvisionerToTest.Set(provisioner)
+		for _, testFunc := range channelTestMap[provisioner] {
+			funcName := runtime.FuncForPC(reflect.ValueOf(testFunc).Pointer()).Name()
+			baseFuncName := GetBaseFuncName(funcName)
+			t.Logf("Running %q with %q ClusterChannelProvisioner", baseFuncName, provisioner)
+			t.Run(baseFuncName, testFunc)
```

Comment:
Is this immediate? In the sense that as soon as this is called, does the test's Setup() get called? 

If it does not happen immediately, I'm worried that ClusterChannelProvisionerToTest may get overwritten before Setup() is called (which reads and caches that value).

Reply:
According to the document, "It runs f in a separate goroutine and blocks until f returns or calls t.Parallel to become a parallel test.". 

`t.Parallel()` is called at the end of `Setup()`, so all setup process will be finished before ClusterChannelProvisionerToTest being overwritten.